arangod/Aql/SimpleModifier.h

```diff
@@ -103,56 +111,60 @@ class SimpleModifier {
   explicit SimpleModifier(ModificationExecutorInfos& infos)
       : _infos(infos),
         _completion(infos),
-        _results(Result(), infos._options),
-        _resultsIterator(VPackArrayIterator::Empty{}),
-        _batchSize(ExecutionBlock::DefaultBatchSize) {}
-  ~SimpleModifier() = default;
+        _batchSize(ExecutionBlock::DefaultBatchSize),
+        _results(NoResult{}) {
+    TRI_ASSERT(_infos.engine() != nullptr);
+  }
+
+  virtual ~SimpleModifier() = default;
 
   void reset();
 
-  Result accumulate(InputAqlItemRow& row);
-  void transact(transaction::Methods& trx);
+  void accumulate(InputAqlItemRow& row);
+  [[nodiscard]] ExecutionState transact(transaction::Methods& trx);
+
+  void checkException() const;
+  void resetResult() noexcept;
 
   // The two numbers below may not be the same: Operations
   // can skip or ignore documents.
 
   // The number of operations
-  size_t nrOfOperations() const;
+  [[nodiscard]] size_t nrOfOperations() const;
   // The number of documents in the accumulator
-  size_t nrOfDocuments() const;
+  [[nodiscard]] size_t nrOfDocuments() const;
   // The number of entries in the results slice
-  size_t nrOfResults() const;
+  [[nodiscard]] [[maybe_unused]] size_t nrOfResults() const;
+
+  [[nodiscard]] size_t nrOfErrors() const;
 
-  size_t nrOfErrors() const;
+  [[nodiscard]] size_t nrOfWritesExecuted() const;
+  [[nodiscard]] size_t nrOfWritesIgnored() const;
 
-  size_t nrOfWritesExecuted() const;
-  size_t nrOfWritesIgnored() const;
+  [[nodiscard]] ModificationExecutorInfos& getInfos() const noexcept;
+  [[nodiscard]] size_t getBatchSize() const noexcept;
 
-  ModificationExecutorInfos& getInfos() const noexcept;
-  size_t getBatchSize() const noexcept;
+  bool hasResultOrException() const noexcept;
+  bool hasNoResultOrOperationPending() const noexcept;
 
  private:
-  bool resultAvailable() const;
-  VPackArrayIterator getResultsIterator() const;
+  [[nodiscard]] bool resultAvailable() const;
+  [[nodiscard]] VPackArrayIterator getResultsIterator() const;
 
   ModificationExecutorInfos& _infos;
   ModifierCompletion _completion;
 
   std::vector<ModOp> _operations;
   ModificationExecutorAccumulator _accumulator;
 
-  OperationResult _results;
-
-  std::vector<ModOp>::const_iterator _operationsIterator;
-  VPackArrayIterator _resultsIterator;
-
   size_t const _batchSize;
+
```

Comment:
As @jsteemann mentioned above:
```suggestion
  ///@brief mutex that protects the _results.
```

Reply:
added the comment here